providers/implementations/exchange/ecdh_exch.c

```diff
@@ -0,0 +1,224 @@
+        return 0;
+    EC_KEY_free(pecdhctx->peerk);
+    pecdhctx->peerk = vecdh;
+    return 1;
+}
+
+static
+void ecdh_freectx(void *vpecdhctx)
+{
+    PROV_ECDH_CTX *pecdhctx = (PROV_ECDH_CTX *)vpecdhctx;
+
+    EC_KEY_free(pecdhctx->k);
+    EC_KEY_free(pecdhctx->peerk);
+
+    OPENSSL_free(pecdhctx);
+}
+
+static
+void *ecdh_dupctx(void *vpecdhctx)
+{
+    PROV_ECDH_CTX *srcctx = (PROV_ECDH_CTX *)vpecdhctx;
+    PROV_ECDH_CTX *dstctx;
+
+    dstctx = OPENSSL_zalloc(sizeof(*srcctx));
+    if (dstctx == NULL)
+        return NULL;
+
+    *dstctx = *srcctx;
+    if (dstctx->k != NULL && !EC_KEY_up_ref(dstctx->k)) {
+        OPENSSL_free(dstctx);
+        return NULL;
+    }
+
+    if (dstctx->peerk != NULL && !EC_KEY_up_ref(dstctx->peerk)) {
+        EC_KEY_free(dstctx->k);
+        OPENSSL_free(dstctx);
+        return NULL;
+    }
+
+    return dstctx;
+}
+
+static
+int ecdh_set_ctx_params(void *vpecdhctx, const OSSL_PARAM params[])
+{
+    PROV_ECDH_CTX *pecdhctx = (PROV_ECDH_CTX *)vpecdhctx;
+    const OSSL_PARAM *p;
+    unsigned int pad;
+
+    if (pecdhctx == NULL || params == NULL)
+        return 0;
+
+#if 0
+    p = OSSL_PARAM_locate_const(params, OSSL_EXCHANGE_PARAM_PAD);
+    if (p == NULL || !OSSL_PARAM_get_uint(p, &pad))
+        return 0;
+    pdhctx->pad = pad ? 1 : 0;
+    return 1;
+#else
+    /* TODO(romen): this is just a stub */
+    (void)p; (void)pad;
+    return 0;
+#endif
+}
+
+static const OSSL_PARAM known_settable_ctx_params[] = {
+    /*
+     * TODO(romen): see crypto/ec/ec_pmeth.c:233
+     *
+     * We should support:
+     *   - parameters for cofactor ECDH (maybe as a separate implementation?)
+     *   - parameters for pkey_ec_kdf_derive
+     */
+#if 0
+    OSSL_PARAM_int(OSSL_EXCHANGE_PARAM_PAD, NULL),
+#endif
+    OSSL_PARAM_END
+};
+
+static
+const OSSL_PARAM *ecdh_settable_ctx_params(void)
+{
+    return known_settable_ctx_params;
+}
+
+static ossl_inline
+size_t ecdh_size(const EC_KEY *k)
+{
+    size_t degree = 0;
+    const EC_GROUP *group;
+
+    if (k == NULL
+            || (group = EC_KEY_get0_group(k)) == NULL)
+        return 0;
+
+    degree = EC_GROUP_get_degree(group);
+
+    return (degree + 7) / 8;
+}
+
+static
+int ecdh_derive(void *vpecdhctx, unsigned char *secret, size_t *secretlen,
+                size_t outlen)
+{
+    PROV_ECDH_CTX *pecdhctx = (PROV_ECDH_CTX *)vpecdhctx;
+    int ret;
+    size_t ecdhsize;
+    const EC_POINT *ppubkey = NULL;
+
+    /* TODO(3.0): Add errors to stack */
+    if (pecdhctx->k == NULL || pecdhctx->peerk == NULL)
+        return 0; /* EC_R_KEYS_NOT_SET */
+
+    ecdhsize = ecdh_size(pecdhctx->k);
+    if (secret == NULL) {
+        *secretlen = ecdhsize;
+        return 1;
+    }
+
+#if 0
+    /* TODO(3.0): pkey_ec_derive() contains the following comment, altough its
+     *            claim about "is truncated" seems misleading:
+     *            I did not test this, but I see the risk for OOB writes with
+     *            this API.
```

Comment:
I did not dive into the KDF handling, but at least on first impressions it does look to me like `ECDH_compute_key` does handle the case where outlen leads to a truncation.

Reply:
Yes, this is a note to self for future debugging sessions once I manage to get things at least partially working.